src/gt4py/next/ffront/stages.py

```diff
@@ -42,3 +89,103 @@ class PastClosure:
     grid_type: Optional[common.GridType]
     args: tuple[Any, ...]
     kwargs: dict[str, Any]
+
+
+# TODO(ricoh): This type seems to not really catch the relevant types
+#   which leads to the ignores below
+Hasher_T: typing.TypeAlias = eve.extended_typing.HashlibAlgorithm
+
+
+def cache_key(obj: Any, algorithm: Optional[str | Hasher_T] = None) -> str:
+    hasher: Hasher_T
+    if not algorithm:
+        hasher = xxhash.xxh64()  # type: ignore[assignment] # see todo above
+    elif isinstance(algorithm, str):
+        hasher = hashlib.new(algorithm)  # type: ignore[assignment] # see todo above
+    else:
+        hasher = algorithm
+
+    update_cache_key(obj, hasher)
+    return hasher.hexdigest()
+
+
+@functools.singledispatch
+def update_cache_key(obj: Any, hasher: Hasher_T) -> None:
+    # the following is to avoid circular dependencies
+    if hasattr(obj, "backend"):  # assume it is a decorator wrapper
+        update_cache_key_fielop(obj, hasher)
+    else:
+        hasher.update(str(obj).encode())
+
+
+@update_cache_key.register(FieldOperatorDefinition)
+@update_cache_key.register(FoastOperatorDefinition)
+@update_cache_key.register(FoastWithTypes)
+@update_cache_key.register(FoastClosure)
+@update_cache_key.register(ProgramDefinition)
+@update_cache_key.register(PastProgramDefinition)
+@update_cache_key.register(PastClosure)
+def update_cache_key_stages(obj: Any, hasher: Hasher_T) -> None:
+    update_cache_key(obj.__class__, hasher)
+    for field in dataclasses.fields(obj):
+        update_cache_key(getattr(obj, field.name), hasher)
+
+
+@update_cache_key.register
+def update_cache_key_str(obj: str, hasher: Hasher_T) -> None:
+    hasher.update(str(obj).encode())
+
+
+@update_cache_key.register(int)
+@update_cache_key.register(bool)
+@update_cache_key.register(float)
+def update_cache_key_builtins(
+    obj: None,
+    hasher: Hasher_T,
+) -> None:
+    hasher.update(str(obj).encode())
+
+
+@update_cache_key.register
+def update_cache_key_func(obj: types.FunctionType, hasher: Hasher_T) -> None:
+    sourcedef = source_utils.SourceDefinition.from_function(obj)
+    for item in sourcedef:
+        update_cache_key(item, hasher)
+
+
+@update_cache_key.register
+def update_cache_key_dict(obj: dict, hasher: Hasher_T) -> None:
+    for key, value in obj.items():
+        update_cache_key(key, hasher)
+        update_cache_key(value, hasher)
+
+
+@update_cache_key.register
+def update_cache_key_type(obj: type, hasher: Hasher_T) -> None:
+    hasher.update(obj.__name__.encode())
+
+
+@update_cache_key.register
+def update_cache_key_sequence(obj: collections.abc.Iterable, hasher: Hasher_T) -> None:
+    for item in obj:
+        update_cache_key(item, hasher)
+
+
+@update_cache_key.register
+def update_cache_key_foast(obj: foast.LocatedNode, hasher: Hasher_T) -> None:
+    update_cache_key(obj.location, hasher)
+    update_cache_key(str(obj), hasher)
+
+
+# not registered to avoid circular dependencies
```

Comment:
I feel like there should be a better way to avoid any circular dependency here but to be honest I don't fully understand the comments or the code. Could you elaborate a bit why this is needed?

Reply:
Never mind. The correct solution is to register the dispatch for `decorator.Program` and `decorator.FieldOperator` inside `decorator`.

The problem seemed to be that `decorator` imports `stages`, so specializing on `decorator` classes in `stages` would require a circular dependency. That would be a problem when not using `singledispatch`.